### PR TITLE
C-0052 - check access to cloud provider meta data API

### PR DESCRIPTION
**Pull request by kooomix**
## Describe your changes
Added "ProviderMetaDataAPIAccess" entry - if true, cloud provider meta data API has access from Node.

References:

**aws API metadata:**
https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
url to check :169.254.169.254/latest/meta-data/local-hostname

**Azure:**
https://learn.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=windows
url to check: http://169.254.169.254/metadata/instance?api-version=2021-02-01
"headers": {"Metadata": "true"}

**Google:**
https://cloud.google.com/compute/docs/metadata/querying-metadata
url to check: http://169.254.169.254/computeMetadata/v1/?alt=json&recursive=true
"headers": {"Metadata-Flavor": "Google"}

## Checklist before requesting a review
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests.
- [x] New and existing unit tests pass locally with my changes

**Please open the PR against the `dev` branch (Unless the PR contains only documentation changes)**

### CODE REVIEW

File: sensor/cloudProvider.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+	if err == nil && res.StatusCode == 200 {
+		return true
+	}
+
+	res, err = httputils.HttpGet(client, "http://169.254.169.254/latest/meta-data/local-hostname", nil)
+
+	if err == nil && res.StatusCode == 200 {
+		return true
+	}
```
Comment (shm12) [inline sensor/cloudProvider.go:45]:
```suggestion
	requests := []struct {
		url     string
		headers map[string]string
	}{
		{
			"http://169.254.169.254/computeMetadata/v1/?alt=json&recursive=true",
			map[string]string{"Metadata-Flavor": "Google"},
		},
		{
			"http://169.254.169.254/metadata/instance?api-version=2021-02-01",
			map[string]string{"Metadata": "true"},
		},
		{
			"http://169.254.169.254/latest/meta-data/local-hostname",
			map[string]string{},
		},
	}

	for _,req := range requests {
		res, err := httputils.HttpGet(client, req.url, req.headers)

		if err == nil && res.StatusCode == 200 {
			return true
		}	
	}
```

File: sensor/cloudProvider.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+	return false
+
+}
```
Comment (shm12) [inline sensor/cloudProvider.go:49]:
```suggestion
var MetaDataAPIRequests = []struct {
	url     string
	headers map[string]string
} 	{
	{
		"http://169.254.169.254/computeMetadata/v1/?alt=json&recursive=true",
		map[string]string{"Metadata-Flavor": "Google"},
	},
	{
		"http://169.254.169.254/metadata/instance?api-version=2021-02-01",
		map[string]string{"Metadata": "true"},
	},
	{
		"http://169.254.169.254/latest/meta-data/local-hostname",
		map[string]string{},
	},
}

func hasMetaDataAPIAccess() bool {
	client := &http.Client{}

	for _,req := range MetaDataAPIRequests {
		res, err := httputils.HttpGet(client, req.url, req.headers)

		if err == nil && res.StatusCode == 200 {
			return true
		}	
	}

	return false

}
```
And then you can unitest it

**Review by shm12 — CHANGES_REQUESTED** [review]
Please add unitests

File: sensor/cloudProvider.go (hunk 2 above)
Comment (kooomix) [reply sensor/cloudProvider.go:49]: What is it that you think needed to unitest here? 
All request will anyway return false (no matter what the params are)...

Comment (shm12) [reply sensor/cloudProvider.go:49]: Everything needs unitets, so the coverage report will look good :)

File: sensor/cloudProvider.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	ret := CloudProviderInfo{}
+
+	ret.ProviderMetaDataAPIAccess = hasMetaDataAPIAccess()
+
+	return &ret, nil
+}
+
+// hasMetaDataAPIAccess - checks if there is an access to cloud provider meta data
+func hasMetaDataAPIAccess() bool {
+	client := &http.Client{}
```
Comment (Bezbran) [inline sensor/cloudProvider.go:46]: In golang http client is an heavy resource.
We should create a global one (singleton) and use it.

**Review by Bezbran — CHANGES_REQUESTED** [review]

File: sensor/cloudProvider.go (hunk 3 above)
Comment (kooomix) [reply sensor/cloudProvider.go:46]: An http client object is being instantiated only once (in the whole host-scanner).
Are you offering to use a global for cases we will want to re-use it in the future?

File: sensor/internal/utils/globals.go [hunk 4]
```diff
@@ -1,7 +1,29 @@
...
+
 var (
 	// Where the host sensor is expecting host fs to be mounted.
 	// Defined as var for testing purposes only
 	HostFileSystemDefaultLocation = "/host_fs"
+
+	// global http.client instance to reduce object resource overuse.
+	httpClient *http.Client
+
+	lock = &sync.Mutex{}
```
Comment (shm12) [inline sensor/internal/utils/globals.go:16]: Please give a more meaningful name to the `lock` object

File: sensor/cloudProvider_test.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	defer httpmock.DeactivateAndReset()
+
+	// Exact URL match
+	httpmock.RegisterResponder("GET", "http://www.mygoodurl.com",
+		httpmock.NewStringResponder(200, `[{"id": 1, "name": "My Good URL"}]`))
+
+	MetaDataAPIRequests = []struct {
+		url     string
+		headers map[string]string
+	}{
```
Comment (shm12) [inline sensor/cloudProvider_test.go:21]: Consider extracting the anonymous struct to named struct, so you don't have to redeclare the struct every time you create it

Comment (kooomix) [reply sensor/cloudProvider_test.go]: done

File: sensor/internal/utils/globals.go (hunk 4 above)
Comment (kooomix) [reply sensor/internal/utils/globals.go]: done

**Review by shm12 — APPROVED** [review]

**Review by Bezbran — APPROVED** [review]

**Pull request merged**